src/result/mod.rs

```diff
@@ -0,0 +1,123 @@
+use std::convert::From;
+use std::fmt::{Debug, Error};
+use std::{fmt, io};
+
+use crate::result::decoding_error::DecodingError;
+use crate::result::encoding_error::EncodingError;
+use crate::result::illegal_operation::IllegalOperation;
+use crate::result::incomplete::IncompleteError;
+use io_error::IoError;
+use position::Position;
+use thiserror::Error;
+
+pub mod decoding_error;
+pub mod encoding_error;
+pub mod illegal_operation;
+pub mod incomplete;
+pub mod io_error;
+pub mod position;
+
+/// A unified Result type representing the outcome of method calls that may fail.
+pub type IonResult<T> = Result<T, IonError>;
+
+/// Represents the different types of high-level failures that might occur when reading Ion data.
+#[derive(Clone, Debug, Error, PartialEq)]
```

Comment:
🗺️ Now that we're using wrapped types that implement all the traits we need, the top-level `IonError` type can derive its implementations of `PartialEq` and `Clone`.